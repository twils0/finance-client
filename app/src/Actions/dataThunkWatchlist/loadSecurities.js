import axios from 'axios';
import raven from 'raven-js';

import { requestStatusTypes, URLs, axiosConfig } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsWatchlist';
import { setWatchlistStatus, setSecuritiesAll } from '../dataActionsWatchlist';
import requestLogout from '../dataThunkAuth/requestLogout';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';

import handleErrorCatch from '../../handleErrorCatch';

import { demo } from '../../../../mode.config.json';
import demoLoadSecurities from '../../../../demoConfig/demoLoadSecurities';

const loadSecurities = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.watchlist.status[statusNames.SECURITIES].status !== requestStatusTypes.LOADING) {
    dispatch(
      setWatchlistStatus({
        id: statusNames.SECURITIES,
        status: requestStatusTypes.LOADING,
      }),
    );

    let { user } = state.data.aws;
    let securities = null;

    if (demo) {
      securities = demoLoadSecurities;
    }

    try {
      if (!user) {
        ({ user } = await dispatch(requestAWSUser()));
      }

      if (!demo) {
        const idToken = user.signInUserSession.idToken.jwtToken;
        const accessToken = user.signInUserSession.accessToken.jwtToken;

        const response = await axios.get(URLs.USERS, {
          headers: {
            Authorization: idToken,
          },
          params: {
            accessToken,
          },
          ...axiosConfig.DB,
        });

        ({ securities } = response.data.body);
      }
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(
        setWatchlistStatus({
          id: statusNames.SECURITIES,
          status: requestStatusTypes.ERROR,
        }),
      );

      if (typeof error === 'object' && error && error.code === 'NotAuthorizedException') {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'loadSecurities',
        });
      }

      return Promise.reject(error);
    }

    if (securities.list.length > 0) {
      dispatch(setSecuritiesAll({ securities }));
    }

    dispatch(
      setWatchlistStatus({
        id: statusNames.SECURITIES,
        status: requestStatusTypes.SUCCESS,
      }),
    );
  }

  return null;
};

export default loadSecurities;
