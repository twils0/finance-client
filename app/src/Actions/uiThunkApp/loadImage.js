import { requestStatusTypes } from '../../Constants/universalConstants';

import { setImageStatus } from '../uiActionsApp';

const loadImage = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'id')) {
    throw new Error(`Please enter a value for the 'id' key - ${JSON.stringify(payload)}`);
  }

  return (dispatch, getState) => {
    const state = getState();
    const { id } = payload;

    if (state.ui.app.images[id] && state.ui.app.images[id].status !== requestStatusTypes.LOADING) {
      dispatch(setImageStatus({ id, status: requestStatusTypes.LOADING }));

      const Images = new Image();

      Images.addEventListener('load', () => dispatch(setImageStatus({ id, status: requestStatusTypes.SUCCESS })));
      Images.addEventListener('error', () => dispatch(setImageStatus({ id, status: requestStatusTypes.ERROR })));

      Images.src = state.ui.app.images[id].src;
    }

    return null;
  };
};

export default loadImage;
