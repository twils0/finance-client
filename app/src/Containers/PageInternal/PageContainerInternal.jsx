import React from 'react';
import PropTypes from 'prop-types';
import { withRouter, Route, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import Loadable from 'react-loadable';
import { BeatLoader } from 'halogenium';

import theme from '../../themes';
import { store } from '../../App';

import { pathNames, requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames as statusNamesAuth } from '../../Constants/dataConstantsAuth';
import { statusNames as statusNamesWatchlist } from '../../Constants/dataConstantsWatchlist';
import { imageNames } from '../../Constants/uiConstantsApp';
import { pageBodyNames } from '../../Constants/uiConstantsInternal';
import { setAuthStatus, setRedirectURL } from '../../Actions/dataActionsAuth';
import setSession from '../../Actions/dataThunkAuth/setSession';
import loadImage from '../../Actions/uiThunkApp/loadImage';
import { setSecuritiesCurrent } from '../../Actions/dataActionsWatchlist';
import loadSecurities from '../../Actions/dataThunkWatchlist/loadSecurities';
import requestUpdateSecurities from '../../Actions/dataThunkWatchlist/requestUpdateSecurities';
import { setCurrentPageBody } from '../../Actions/uiActionsInternal';

import Loading from '../../Components/Loading';
import FlexColumn from '../../Components/Flex/FlexColumn';
import Page from '../../Components/Page';
import TopMenuContainerInternal from './TopMenuContainerInternal/TopMenuContainerInternal';

const loadableState = store.getState();
const { menu } = loadableState.ui.internal.utilities;

const LoadableWatchlist = Loadable({
  loader: () => import('./PageBodyContainerWatchlist/PageBodyContainerWatchlist'),
  loading: () => <Loading menuHeight={menu.height + menu.heightBuffer} />,
  delay: 0,
});

const LoadableAccount = Loadable({
  loader: () => import('./PageBodyContainerAccount/PageBodyContainerAccount'),
  loading: () => <Loading menuHeight={menu.height + menu.heightBuffer} />,
  delay: 0,
});

class PageContainerInternal extends React.Component {
  constructor(props) {
    super(props);

    const {
      handleSession,
      redirectURL,
      handleAuthStatus,
      statusAuth,
      statusexampleHeader,
      statusAWS,
      statusSecurities,
      securities,
      match,
      currentPageBody,
    } = this.props;
    const { path, url, params } = match;
    const { securityId } = params;

    this.initSessionAuth = handleSession();

    if (redirectURL !== url) {
      this.setRedirect(match);
    }

    if (statusAuth[statusNamesAuth.LOGOUT].status !== requestStatusTypes.IDLE) {
      handleAuthStatus({ id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE });
    }
    if (statusAuth[statusNamesAuth.DELETE_ACCOUNT].status !== requestStatusTypes.IDLE) {
      handleAuthStatus({ id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE });
    }

    if (statusexampleHeader === requestStatusTypes.IDLE) {
      this.props.handleImage({ id: imageNames.EXAMPLE_HEADER });
    }

    if (statusAWS === requestStatusTypes.SUCCESS && statusSecurities === requestStatusTypes.IDLE) {
      this.props.handleSecurities();
    }

    this.setCurrentSecurity(path, securityId, securities);

    this.setCurrentPageBody(path, currentPageBody);
  }

  componentDidMount() {
    if (this.initSessionAuth) {
      this.initSessionAuth = false;
    }
  }

  componentWillReceiveProps(nextProps) {
    const {
      statusAWS, redirectURL, statusSecurities, match, securities, currentPageBody,
    } = nextProps;
    const { path, url, params } = match;
    const { securityId } = params;

    if (statusAWS === requestStatusTypes.SUCCESS && statusSecurities === requestStatusTypes.IDLE) {
      this.props.handleSecurities();
    }

    if (this.props.match.url !== url && redirectURL !== url) {
      this.setRedirect(match);
    }

    this.setCurrentSecurity(path, securityId, securities);

    this.setCurrentPageBody(path, currentPageBody);
  }

  componentWillUpdate() {
    this.props.handleSession();
  }

  setRedirect = (match) => {
    const { path, url } = match;

    if (
      path === pathNames.WATCHLIST_SECURITY_ID
      || path === pathNames.WATCHLIST
      || path === pathNames.ACCOUNT
    ) {
      this.props.handleRedirectURL({ redirectURL: url });
    }
  };

  setCurrentSecurity = (path, securityId, securities) => {
    if (path === pathNames.WATCHLIST || path === pathNames.WATCHLIST_SECURITY_ID) {
      if (
        securityId
        && securityId !== ':securityId'
        && securities
        && Object.keys(securities).indexOf(securityId) > -1
      ) {
        if (securities && securityId !== securities.current) {
          const { handleSecuritiesCurrent, handleUpdateSecurities } = this.props;

          handleSecuritiesCurrent({ current: securityId });
          handleUpdateSecurities();
        }
      } else if (securities && securities.current) {
        this.props.history.replace(`${pathNames.WATCHLIST}/${securities.current}`);
      }
    }
  };

  setCurrentPageBody = (path, currentPageBody) => {
    if (path === pathNames.WATCHLIST_SECURITY_ID && currentPageBody !== pageBodyNames.WATCHLIST) {
      this.props.handleCurrentPageBody({ current: pageBodyNames.WATCHLIST });
    }
    if (path === pathNames.ACCOUNT && currentPageBody !== pageBodyNames.ACCOUNT) {
      this.props.handleCurrentPageBody({ current: pageBodyNames.ACCOUNT });
    }
  };

  render() {
    const {
      authenticated, statusFonts, statusexampleHeader, statusAWS, statusAuth,
    } = this.props;

    if (!this.initSessionAuth && !authenticated) {
      return <Redirect to={pathNames.LOGIN} />;
    }

    if (
      statusFonts !== requestStatusTypes.SUCCESS
      || statusexampleHeader !== requestStatusTypes.SUCCESS
      || statusAWS === requestStatusTypes.LOADING
      || statusAuth[statusNamesAuth.LOGOUT].status === requestStatusTypes.LOADING
      || statusAuth[statusNamesAuth.LOGOUT].status === requestStatusTypes.SUCCESS
      || statusAuth[statusNamesAuth.DELETE_ACCOUNT].status === requestStatusTypes.LOADING
      || statusAuth[statusNamesAuth.DELETE_ACCOUNT].status === requestStatusTypes.SUCCESS
    ) {
      return (
        <FlexColumn width="100%" height="100%">
          <BeatLoader color={theme.color.secondary.default} />
        </FlexColumn>
      );
    }

    return (
      <Page id="pageInternal" noBackgroundImage>
        <TopMenuContainerInternal />
        <Route path={pathNames.WATCHLIST_SECURITY_ID} component={LoadableWatchlist} />
        <Route path={pathNames.ACCOUNT} component={LoadableAccount} />
      </Page>
    );
  }
}

PageContainerInternal.propTypes = {
  match: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  redirectURL: PropTypes.string.isRequired,
  statusAuth: PropTypes.object.isRequired,
  statusexampleHeader: PropTypes.string.isRequired,
  securities: PropTypes.object.isRequired,
  currentPageBody: PropTypes.string.isRequired,
  authenticated: PropTypes.bool.isRequired,
  statusFonts: PropTypes.string.isRequired,
  statusAWS: PropTypes.string.isRequired,
  statusSecurities: PropTypes.string.isRequired,
  handleSession: PropTypes.func.isRequired,
  handleAuthStatus: PropTypes.func.isRequired,
  handleSecuritiesCurrent: PropTypes.func.isRequired,
  handleUpdateSecurities: PropTypes.func.isRequired,
  handleCurrentPageBody: PropTypes.func.isRequired,
  handleImage: PropTypes.func.isRequired,
  handleSecurities: PropTypes.func.isRequired,
  handleRedirectURL: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  match: ownProps.match,
  history: ownProps.history,
  redirectURL: state.data.auth.redirectURL,
  statusAuth: state.data.auth.status,
  statusexampleHeader: state.ui.app.images.exampleHeader.status,
  securities: state.data.watchlist.securities,
  currentPageBody: state.ui.internal.pageBodies.current,
  authenticated: state.data.auth.authenticated,
  statusFonts: state.ui.app.fonts.status,
  statusAWS: state.data.aws.status,
  statusSecurities: state.data.watchlist.status[statusNamesWatchlist.SECURITIES].status,
});

const mapDispatchToProps = dispatch => ({
  handleSession: () => dispatch(setSession()),
  handleAuthStatus: payload => dispatch(setAuthStatus(payload)),
  handleSecuritiesCurrent: payload => dispatch(setSecuritiesCurrent(payload)),
  handleUpdateSecurities: payload => dispatch(requestUpdateSecurities(payload)),
  handleCurrentPageBody: payload => dispatch(setCurrentPageBody(payload)),
  handleImage: id => dispatch(loadImage(id)),
  handleSecurities: () => dispatch(loadSecurities()),
  handleRedirectURL: payload => dispatch(setRedirectURL(payload)),
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(PageContainerInternal),
);
