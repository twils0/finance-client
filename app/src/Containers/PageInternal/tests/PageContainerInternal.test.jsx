import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { store as importedStore } from '../../../App';

import { initialState as initialStateAuth } from '../../../Reducers/dataReducersAuth';
import { initialState as initialStateAWS } from '../../../Reducers/dataReducersAWS';
import { initialState as initialStateWatchlist } from '../../../Reducers/dataReducersWatchlist';
import { initialState as initialStateUIApp } from '../../../Reducers/uiReducersApp';
import { initialState as initialStateUIInternal } from '../../../Reducers/uiReducersInternal';
import { requestStatusTypes, pathNames } from '../../../Constants/universalConstants';
import {
  actionTypes as actionTypesAuth,
  statusNames as statusNamesAuth,
} from '../../../Constants/dataConstantsAuth';
import { actionTypes as actionTypesWatchlist } from '../../../Constants/dataConstantsWatchlist';
import { imageNames } from '../../../Constants/uiConstantsApp';
import {
  actionTypes as actionTypesUIInternal,
  pageBodyNames,
} from '../../../Constants/uiConstantsInternal';
import loadSecurities from '../../../Actions/dataThunkWatchlist/loadSecurities';
import requestUpdateSecurities from '../../../Actions/dataThunkWatchlist/requestUpdateSecurities';
import loadImage from '../../../Actions/uiThunkApp/loadImage';

import PageContainerInternal from '../PageContainerInternal';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

importedStore.getState = jest.fn(() => ({ ui: { internal: initialStateUIInternal } }));

jest.mock('../../../Actions/uiThunkApp/loadImage', () => jest.fn(() => () => null));
jest.mock('../../../Actions/dataThunkWatchlist/loadSecurities', () => jest.fn());
jest.mock('../../../Actions/dataThunkWatchlist/requestUpdateSecurities', () => jest.fn());

const history = {
  replace: jest.fn(),
};

global.window = {
  innerHeight: 500,
};
global.document = {};

const shallowComponent = (store, path, url, params) =>
  shallow(<PageContainerInternal />, {
    context: {
      store,
      router: {
        history,
        route: {
          location: {},
          match: { path, url, params: params || {} },
        },
      },
    },
  })
    .dive()
    .dive()
    .dive();

const testId0 = 'testId0';
const testId1 = 'testId1';
const testId2 = 'testId2';
const testId3 = 'testId3';

const securities = {
  current: testId0,
  list: [testId0, testId1, testId2],
  [testId0]: {
    id: testId0,
    name: 'testName0',
    category: 'testCategory0',
    exchange: 'testExchange0',
    tickerCusip: 'testTickerCusip0',
    data: {
      description: 'testDescription0',
    },
  },
  [testId1]: {
    id: testId1,
    name: 'testName1',
    category: 'testCategory1',
    exchange: 'testExchange1',
    tickerCusip: 'testTickerCusip1',
    data: {
      description: 'testDescription1',
    },
  },
  [testId2]: {
    id: testId2,
    name: 'testName2',
    category: 'testCategory2',
    exchange: 'testExchange2',
    tickerCusip: 'testTickerCusip2',
    data: {
      description: 'testDescription2',
    },
  },
};

describe('Containers', () => {
  describe('PageInternal', () => {
    describe('PageContainerInternal', () => {
      afterEach(() => {
        importedStore.getState.mockReset();
        loadImage.mockReset();
        loadSecurities.mockReset();
        requestUpdateSecurities.mockReset();
        history.replace.mockReset();
      });

      it('shallow renders correctly, with logout and delete account statuses set to success, watchlist path, and no securityId param', async () => {
        const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
        const stateBeforeWatchlist = JSON.parse(JSON.stringify(initialStateWatchlist));
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));

        stateBeforeAWS.status = requestStatusTypes.SUCCESS;

        stateBeforeWatchlist.securities = securities;

        stateBeforeUIApp.fonts.status = requestStatusTypes.SUCCESS;

        const store = mockStore({
          data: {
            auth: initialStateAuth,
            aws: stateBeforeAWS,
            watchlist: stateBeforeWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: initialStateUIInternal,
          },
        });

        const url = `${pathNames.WATCHLIST}/123`;
        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: url },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE },
          },
        ];

        loadImage.mockReturnValue(() => null);
        loadSecurities.mockReturnValue(() => null);

        const wrapper = shallowComponent(store, pathNames.WATCHLIST, url);

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(loadImage).toBeCalledWith({ id: imageNames.EXAMPLE_HEADER });
        expect(loadSecurities).toHaveBeenCalledTimes(1);
        expect(requestUpdateSecurities).not.toBeCalled();
        expect(history.replace).toBeCalledWith(`${pathNames.WATCHLIST}/${securities.current}`);
        expect(history.replace).toHaveBeenCalledTimes(1);
      });

      it('shallow renders correctly, with componentDidMount, componentWillReceiveProps, watchlist path, securityId param, and account page body', async () => {
        const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
        const stateBeforeWatchlist = JSON.parse(JSON.stringify(initialStateWatchlist));
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
        const stateBeforeUIInternal = JSON.parse(JSON.stringify(initialStateUIInternal));

        stateBeforeAWS.status = requestStatusTypes.SUCCESS;

        stateBeforeWatchlist.securities = securities;

        stateBeforeUIApp.images.exampleHeader.status = requestStatusTypes.SUCCESS;
        stateBeforeUIApp.fonts.status = requestStatusTypes.SUCCESS;

        stateBeforeUIInternal.pageBodies.current = pageBodyNames.ACCOUNT;

        const store = mockStore({
          data: {
            auth: initialStateAuth,
            aws: stateBeforeAWS,
            watchlist: stateBeforeWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: stateBeforeUIInternal,
          },
        });

        const urlStart = pathNames.ACCOUNT;
        const urlNext = `${pathNames.WATCHLIST}/123`;
        const nextProps = {
          match: {
            path: pathNames.WATCHLIST_SECURITY_ID,
            url: urlNext,
            params: {
              securityId: testId1,
            },
          },
          statusAWS: requestStatusTypes.SUCCESS,
          securities,
        };

        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: urlStart },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: urlNext },
          },
          {
            type: actionTypesWatchlist.SET_SECURITIES_CURRENT,
            payload: { current: testId1 },
          },
          {
            type: actionTypesUIInternal.SET_CURRENT_PAGE_BODY,
            payload: { current: pageBodyNames.WATCHLIST },
          },
        ];

        loadSecurities.mockReturnValue(() => null);
        requestUpdateSecurities.mockReturnValue(() => null);

        const wrapper = shallowComponent(store, pathNames.ACCOUNT, urlStart, {
          securityId: testId0,
        });

        const instance = wrapper.instance();

        instance.componentDidMount();
        instance.componentWillReceiveProps(nextProps);

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(loadSecurities).toHaveBeenCalledTimes(1);
        expect(requestUpdateSecurities).toBeCalled();
        expect(history.replace).not.toBeCalled();
      });

      it('shallow renders correctly, when authenticated equals false, with watchlist path and securityId param equal to current security', async () => {
        const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
        const stateBeforeWatchlist = JSON.parse(JSON.stringify(initialStateWatchlist));
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));

        stateBeforeAuth.authenticated = false;

        stateBeforeWatchlist.securities = securities;

        stateBeforeUIApp.images.exampleHeader.status = requestStatusTypes.SUCCESS;
        stateBeforeUIApp.fonts.status = requestStatusTypes.SUCCESS;

        const store = mockStore({
          data: {
            auth: stateBeforeAuth,
            aws: initialStateAWS,
            watchlist: stateBeforeWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: initialStateUIInternal,
          },
        });

        const url = `${pathNames.WATCHLIST}/123`;
        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: url },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE },
          },
        ];

        const wrapper = shallowComponent(store, pathNames.WATCHLIST, url, { securityId: testId0 });

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(history.replace).not.toBeCalled();
        expect(loadImage).not.toBeCalled();
        expect(requestUpdateSecurities).not.toBeCalled();
        expect(history.replace).not.toBeCalled();
      });

      it('shallow renders correctly, with watchlist path and securityId param not equal to current security', async () => {
        const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
        const stateBeforeWatchlist = JSON.parse(JSON.stringify(initialStateWatchlist));
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));

        stateBeforeWatchlist.securities = securities;

        stateBeforeUIApp.images.exampleHeader.status = requestStatusTypes.SUCCESS;
        stateBeforeUIApp.fonts.status = requestStatusTypes.SUCCESS;

        const store = mockStore({
          data: {
            auth: stateBeforeAuth,
            aws: initialStateAWS,
            watchlist: stateBeforeWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: initialStateUIInternal,
          },
        });

        const url = `${pathNames.WATCHLIST}/123`;
        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: url },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesWatchlist.SET_SECURITIES_CURRENT,
            payload: { current: testId1 },
          },
        ];

        requestUpdateSecurities.mockReturnValue(() => null);

        const wrapper = shallowComponent(store, pathNames.WATCHLIST, url, { securityId: testId1 });

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(loadImage).not.toBeCalled();
        expect(requestUpdateSecurities).toBeCalled();
        expect(history.replace).not.toBeCalled();
      });

      it('shallow renders correctly, with watchlist path and no current security', async () => {
        const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
        const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));

        stateBeforeAWS.status = requestStatusTypes.SUCCESS;

        stateBeforeUIApp.images.exampleHeader.status = requestStatusTypes.SUCCESS;
        stateBeforeUIApp.fonts.status = requestStatusTypes.SUCCESS;

        const store = mockStore({
          data: {
            auth: stateBeforeAuth,
            aws: stateBeforeAWS,
            watchlist: initialStateWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: initialStateUIInternal,
          },
        });

        const url = `${pathNames.WATCHLIST}/123`;
        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: url },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE },
          },
        ];

        loadSecurities.mockReturnValue(() => null);

        const wrapper = shallowComponent(store, pathNames.WATCHLIST, url);

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(loadImage).not.toBeCalled();
        expect(loadSecurities).toHaveBeenCalledTimes(1);
        expect(requestUpdateSecurities).not.toBeCalled();
        expect(history.replace).not.toBeCalled();
      });

      it('shallow renders correctly, with watchlist path and securityId param not in securities', async () => {
        const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
        const stateBeforeWatchlist = JSON.parse(JSON.stringify(initialStateWatchlist));
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));

        stateBeforeWatchlist.securities = securities;

        stateBeforeUIApp.images.exampleHeader.status = requestStatusTypes.SUCCESS;
        stateBeforeUIApp.fonts.status = requestStatusTypes.SUCCESS;

        const store = mockStore({
          data: {
            auth: stateBeforeAuth,
            aws: initialStateAWS,
            watchlist: stateBeforeWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: initialStateUIInternal,
          },
        });

        const url = `${pathNames.WATCHLIST}/123`;
        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: url },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE },
          },
        ];

        requestUpdateSecurities.mockReturnValue(() => null);

        const wrapper = shallowComponent(store, pathNames.WATCHLIST, url, { securityId: testId3 });

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(loadImage).not.toBeCalled();
        expect(requestUpdateSecurities).not.toBeCalled();
        expect(history.replace).toBeCalledWith(`${pathNames.WATCHLIST}/${testId0}`);
      });

      it('shallow renders correctly, with account path, fonts status equal to loading, no watchlist securities, and loading indicator', async () => {
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));

        stateBeforeUIApp.images.exampleHeader.status = requestStatusTypes.SUCCESS;
        stateBeforeUIApp.fonts.status = requestStatusTypes.LOADING;

        const store = mockStore({
          data: {
            auth: initialStateAuth,
            aws: initialStateAWS,
            watchlist: initialStateWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: initialStateUIInternal,
          },
        });

        const url = pathNames.ACCOUNT;
        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: url },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.LOGOUT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesAuth.SET_AUTH_STATUS,
            payload: { id: statusNamesAuth.DELETE_ACCOUNT, status: requestStatusTypes.IDLE },
          },
          {
            type: actionTypesUIInternal.SET_CURRENT_PAGE_BODY,
            payload: { current: pageBodyNames.ACCOUNT },
          },
        ];

        const wrapper = shallowComponent(store, pathNames.ACCOUNT, url);

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(loadImage).not.toBeCalled();
        expect(requestUpdateSecurities).not.toBeCalled();
        expect(history.replace).not.toBeCalled();
      });

      it('shallow renders correctly, with watchlist path, securityId equal to current security, fonts, example header, and aws statuses image equal to success, and logout and delete account status equal to idle', async () => {
        const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
        const stateBeforeWatchlist = JSON.parse(JSON.stringify(initialStateWatchlist));
        const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
        const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));

        stateBeforeAuth.status[statusNamesAuth.LOGOUT].status = requestStatusTypes.IDLE;
        stateBeforeAuth.status[statusNamesAuth.DELETE_ACCOUNT].status = requestStatusTypes.IDLE;

        stateBeforeAWS.status = requestStatusTypes.SUCCESS;

        stateBeforeWatchlist.securities = securities;

        stateBeforeUIApp.images.exampleHeader.status = requestStatusTypes.SUCCESS;
        stateBeforeUIApp.fonts.status = requestStatusTypes.SUCCESS;

        const store = mockStore({
          data: {
            auth: stateBeforeAuth,
            aws: stateBeforeAWS,
            watchlist: stateBeforeWatchlist,
          },
          ui: {
            app: stateBeforeUIApp,
            internal: initialStateUIInternal,
          },
        });

        const url = `${pathNames.WATCHLIST}/123`;
        const expectedActions = [
          {
            type: actionTypesAuth.SET_REDIRECT_URL,
            payload: { redirectURL: url },
          },
        ];

        loadSecurities.mockReturnValue(() => null);

        const wrapper = shallowComponent(store, pathNames.WATCHLIST, url, { securityId: testId0 });

        const actions = store.getActions();

        expect(actions).toEqual(expectedActions);
        expect(wrapper).toMatchSnapshot();
        expect(loadImage).not.toBeCalled();
        expect(loadSecurities).toHaveBeenCalledTimes(1);
        expect(requestUpdateSecurities).not.toBeCalled();
        expect(history.replace).not.toBeCalled();
      });
    });
  });
});
