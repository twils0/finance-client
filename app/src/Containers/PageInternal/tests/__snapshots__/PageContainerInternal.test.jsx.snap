// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, when authenticated equals false, with watchlist path and securityId param equal to current security 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, with account path, fonts status equal to loading, no watchlist securities, and loading indicator 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, with componentDidMount, componentWillReceiveProps, watchlist path, securityId param, and account page body 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, with logout and delete account statuses set to success, watchlist path, and no securityId param 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, with watchlist path and no current security 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, with watchlist path and securityId param not equal to current security 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, with watchlist path and securityId param not in securities 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Containers PageInternal PageContainerInternal shallow renders correctly, with watchlist path, securityId equal to current security, fonts, example header, and aws statuses image equal to success, and logout and delete account status equal to idle 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;
